### PR TITLE
Add InsertOrGet method

### PR DESCRIPTION
**Pull request by bvandre**
I have a need for this method in an application I am writing.  I have a struct that I need to add to the tree, but if it is already in the tree I just want to pull that.  Right now I have to do a get, test if the get is nil and then insert.  With this method I can do it all in the same step.

I have added a test for this method to show that it works and to show what I would intend to use it for.

### CODE REVIEW

**Comment by HuKeping** [comment]
Would you mind squash it into one commit, it will be help especially when use `git bisect`

**Comment by bvandre** [comment]
There we go, all squashed.

File: rbtree.go [hunk 1]
```diff
@@ -133,7 +133,7 @@ func (t *Rbtree) rightRotate(x *Node) {
 	x.Parent = y
 }
 
-func (t *Rbtree) insert(z *Node) {
+func (t *Rbtree) insert(z *Node) Item {
```
Comment (HuKeping) [inline rbtree.go]: If we must return a value here, I prefer it be a `Node`.  Let those exported methods to deal with the things like `Item`, WDYT

Comment (bvandre) [inline rbtree.go]: I see your point and I agree with you.  Everything in there is just dealing with Nodes so we should keep it consistent.  I will update it tonight sometime.

**Comment by bvandre** [comment]
Alright got the updated unexported insert to return a Node and the exported InsertOrGet to get the item off that node.  Let me know if you have any other worries.

**Comment by HuKeping** [comment]
LGTM

**Pull request merged**